### PR TITLE
In Rails >= 5, use `optional: true`, not `required: false`

### PR DESCRIPTION
**Pull request by petehamilton**
The `required` option on belongs_to associations has been deprecated,
instead the recommended (and default) approach is to set this to true by
default and disable it using `optional: true`.

The developer can override this default globally, however, so we can't
say that things like `optional: false` are superfluous, or indeed that
`required: true` isn't necessary.

This change therefore only makes the known-to-be-valid correction from
`required: false` to `optional: true` and registers an offence for any other
instances of `required`.

-----------------

I _think_ I've followed all the instructions correctly and this is ready for review. Let me know if the comment/commit is unclear, or if there are any changes which could make this into a better contribution.

* [x] Wrote [good commit messages][1].
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## master (unreleased)
 
+### New features
+
+* [#6596](https://github.com/rubocop-hq/rubocop/pull/6596): Add offence when using `required` option for `belongs_to` associations in Rails >= 5. ([@petehamilton][])
```
Comment (bbatsov) [inline CHANGELOG.md]: I think you forgot to add yourself to the bottom of the changelog.

**Comment by petehamilton** [comment]
@bquorning @bbatsov apologies - didn't realise this had gone red 🤔 

Worth me getting it all passing again and then we can take another pass at merging in?

File: CHANGELOG.md (hunk 1 above)
Comment (petehamilton) [reply CHANGELOG.md]: I don't think so - already there? 😄

**Comment by petehamilton** [comment]
So, looks like I was getting a failure relating to the docs perhaps?

<details>

<summary>Log Trace</summary>
```
Failures:

  1) RuboCop Project default configuration file has a SupportedStyles for all EnforcedStyle and EnforcedStyle is valid
     Failure/Error:
       enforced_styles = config[name]
                         .select { |key, _| key.start_with?('Enforced') }

     NoMethodError:
       private method `select' called for nil:NilClass
     # ./spec/project_spec.rb:47:in `block (4 levels) in <top (required)>'
     # ./spec/project_spec.rb:45:in `each'
     # ./spec/project_spec.rb:45:in `block (3 levels) in <top (required)>'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner/rspec3.rb:37:in `block (3 levels) in run_specs'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/iterator.rb:57:in `block in each'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner.rb:406:in `around_filter'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/iterator.rb:57:in `call'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/iterator.rb:57:in `each'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner/rspec3.rb:26:in `map'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner/rspec3.rb:26:in `block (2 levels) in run_specs'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner/rspec3.rb:25:in `block in run_specs'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner/rspec3.rb:24:in `run_specs'
     # tasks/spec_runner.rake:67:in `run_worker'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner.rb:301:in `block (2 levels) in spawn_workers'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner.rb:296:in `fork'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner.rb:296:in `block in spawn_workers'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner.rb:293:in `times'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner.rb:293:in `spawn_workers'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner.rb:238:in `execute_internal'
     # ./.bundle/gems/test-queue-0.4.2/lib/test_queue/runner.rb:129:in `execute'
     # tasks/spec_runner.rake:27:in `block in run_specs'
     # tasks/spec_runner.rake:41:in `with_encoding'
     # tasks/spec_runner.rake:25:in `run_specs'
     # tasks/spec_runner.rake:112:in `block in <top (required)>'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:273:in `block in execute'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:273:in `each'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:273:in `execute'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:214:in `block in invoke_with_call_chain'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:194:in `invoke_with_call_chain'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:238:in `block in invoke_prerequisites'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:236:in `each'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:236:in `invoke_prerequisites'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:213:in `block in invoke_with_call_chain'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:194:in `invoke_with_call_chain'
     # ./.bundle/gems/rake-12.3.2/lib/rake/task.rb:183:in `invoke'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:160:in `invoke_task'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:116:in `block (2 levels) in top_level'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:116:in `each'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:116:in `block in top_level'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:125:in `run_with_threads'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:110:in `top_level'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:83:in `block in run'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:186:in `standard_exception_handling'
     # ./.bundle/gems/rake-12.3.2/lib/rake/application.rb:80:in `run'
     # ./.bundle/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
     # ./.bundle/binstubs/bundle:104:in `load'
     # ./.bundle/binstubs/bundle:104:in `<main>'
```

</details>

**Comment by bquorning** [comment]
Some of the project_spec examples don’t give much clue about what’s wrong when they fail. I think this one is the most useful one:
```
  3) RuboCop Project default configuration file has configuration for all cops
     Failure/Error: expect(configuration_keys).to match_array(%w[AllCops Rails] + cop_names)

       expected collection contained:  ["AllCops", "Bundler/DuplicatedGem", "Bundler/GemComment", "Bundler/InsecureProtocolSource", "Bundler..., "Style/WhileUntilModifier", "Style/WordArray", "Style/YodaCondition", "Style/ZeroLengthPredicate"]
       actual collection contained:    ["AllCops", "Bundler/DuplicatedGem", "Bundler/GemComment", "Bundler/InsecureProtocolSource", "Bundler..., "Style/WhileUntilModifier", "Style/WordArray", "Style/YodaCondition", "Style/ZeroLengthPredicate"]
       the missing elements were:      ["Rails/BelongsTo"]
     # ./spec/project_spec.rb:24:in `block (3 levels) in <top (required)>'
```

**Comment by petehamilton** [comment]
@bquorning 🙌 - thanks, I definitely should have spotted that. Fix incoming.

**Pull request merged**